packages/angular_devkit/build_optimizer/src/transforms/prefix-classes.ts

```diff
@@ -0,0 +1,179 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import * as ts from 'typescript';
+
+
+export function testPrefixClasses(content: string) {
+  const regexes = [
+    // tslint:disable-next-line:max-line-length
+    /^(var (\S+) = )(\(function \(\) \{\r?\n(?:    (?:\/\*\*| \*|\*\/|\/\/)[^\r?\n]*\r?\n)*    function \2\([^\)]*\) \{\r?\n)/,
+    /^(var (\S+) = )(\(function \(_super\) \{\r?\n    \w*__extends\(\w+, _super\);\r?\n)/,
+  ];
+
+  return regexes.some((regex) => regex.test(content));
+}
+
+export function getPrefixClassesTransformer(): ts.TransformerFactory<ts.SourceFile> {
+  return (context: ts.TransformationContext): ts.Transformer<ts.SourceFile> => {
+    const transformer: ts.Transformer<ts.SourceFile> = (sf: ts.SourceFile) => {
+
+      const pureFunctionComment = '@__PURE__';
+
+      const visitor: ts.Visitor = (node: ts.Node): ts.Node => {
+
+        // Add pure comment to downleveled classes.
+        if (isDownleveledClass(node)) {
+          const varDecl = node as ts.VariableDeclaration;
+          const varDeclInit = varDecl.initializer as ts.Expression;
+
+          // Create a new node with the pure comment before the variable declaration initializer.
+          const newNode = ts.createVariableDeclaration(
+            varDecl.name,
+            undefined,
+            ts.addSyntheticLeadingComment(
+              varDeclInit, ts.SyntaxKind.MultiLineCommentTrivia, pureFunctionComment, false,
+            ),
+          );
+
+          // Replace node with modified one.
+          return ts.visitEachChild(newNode, visitor, context);
+        }
+
+        // Otherwise return node as is.
+        return ts.visitEachChild(node, visitor, context);
+      };
+
+      return ts.visitNode(sf, visitor);
```

Comment:
`ts.visitEachChild` would be slightly more efficient here.  The source file node can't be a down leveled class so no need to check it.